### PR TITLE
Show JSX if PropVal is a React element

### PR DESCRIPTION
**Pull request by sventschui**
Issue:

## What I did

JSX elements within prop values are now rendered completely (in Story Source and Prop Table). The behaviour can be reverted by configuring `showSourceOfProps: false`.

This yields some errors in the source code generation (with indentation).

I think it would be easier to do the JSX source generation outside of React components since there are a lot of "inverted" dependencies. Rendering the parent depends often on whether the child has multiple lines of code or just a single one.

There is some paddingLeft used for indentation  currently what breaks copy & paste behaviour. I think some of the issues outlined could be fixed by using paddingLeft in some cases but I'd rather not do that (because copy & paste).

Would you be open to a diffrent approach on rendering the source code?

image

## How to test

Create a story with a property value of type JSX element.

```
  .addWithInfo('Test', () => {
    const Test = ({ foo, bar, children, arr }) => (<div>
      <div>{foo}</div>
      <div>{bar}</div>
      {arr}
      <div>{children}</div>
    </div>);

    Test.propTypes = {
      foo: PropTypes.node.isRequired,
      bar: PropTypes.node,
      arr: PropTypes.arrayOf(PropTypes.node),
      obj: PropTypes.objectOf(PropTypes.node),
      children: PropTypes.node.isRequired,
    };

    Test.defaultProps = {
      bar: <div>default works in PropTable!</div>,
      arr: null,
      obj: null,
    };

    return (
      <Test foo={<div />} arr={[<Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />]}>
        <Test
          foo={<div>foo <div>bar</div></div>}
          bar={<div>test</div>}
          arr={[<Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />, <Test foo={<div />} />]}
          obj={{ foo: <Test foo={<div />} />, bar: <Test foo={<div />} />, }}
        >
          <div>children</div>
        </Test>
      </Test>
    );
  })
```

### CODE REVIEW

File: addons/info/src/components/PropVal.js [hunk 1]
```diff
@@ -83,9 +87,16 @@ function previewObject(val, maxPropObjectKeys) {
 }
 
 export default function PropVal(props) {
-  const { maxPropObjectKeys, maxPropArrayLength, maxPropStringLength } = props;
+  const {
+    braceWrap: _branceWrap,
```
Comment (danielduan) [inline addons/info/src/components/PropVal.js]: `_braceWrap`?

File: addons/info/src/components/Node.js [hunk 2]
```diff
@@ -26,7 +26,7 @@ function getData(element) {
   }
 
   if (typeof element === 'number') {
-    data.text = String.toString(element);
+    data.text = Number.prototype.toString.call(element);
```
Comment (danielduan) [inline addons/info/src/components/Node.js]: Does this achieve something different than `element + ''`?

File: addons/info/src/components/Node.js [hunk 3]
```diff
@@ -46,6 +46,7 @@ export default function Node(props) {
   const {
     node,
     depth,
+    showSourceOfProps,
```
Comment (danielduan) [inline addons/info/src/components/Node.js]: I wonder if this should get passed as a context instead.

**Review by danielduan — APPROVED** [review]
Looks good! Thanks for the PR!

Can you add an example of this in the kitchen sink example app? You should also take a quick glance at the existing info usages there and see that nothing breaks.

**Comment by danielduan** [comment]
To answer your question on the rendering part, I don't think any of us are married to a specific method. A lot of the decisions were made way back when it was hard to predict the complexity of info.

If you think there's a more logical way to render it, we'd love a PR.

File: addons/info/src/components/PropVal.js (hunk 1 above)
Comment (sventschui) [inline addons/info/src/components/PropVal.js]: nice catch

**Comment by sventschui** [comment]
I'll try to elaborate on a more robust way to render the source. I think the copy & paste use case is big enough issue for me to stop using padding-left as indentation.

I'll add an example to the kitchen sink. Thanks for the heads up.

**Comment by ndelangen** [comment]
@sventschui could you have a look at the merge conflicts? 🙇

**Comment by sventschui** [comment]
@ndelangen I think you can close this PR. This will yield some errors. The source code generation thingy needs to be refactored

**Pull request closed without merging**